File: Sentinel/iotjava/iotrmi/C++/basics/TestClassAdvanced_Stub.cpp
#include <iostream>
#include <string>
#include "TestClassComplete_Stub.hpp"
#include "CallBack.hpp"

using namespace std;

int main(int argc, char *argv[])
{

	int port = 5010;
	const char* address = "localhost";
	//const char* address = "192.168.2.191";	// RPi2
	//const char* skeletonAddress = "128.195.136.170";	// dc-9.calit2.uci.edu
	const char* skeletonAddress = "128.195.204.132";
	const char* callbackAddress = "128.195.204.132";	// dw-2.eecs.uci.edu (this machine)
	//const char* skeletonAddress = "192.168.2.108";	// RPi1
	//const char* callbackAddress = "192.168.2.191";	// RPi2
	int rev = 0;
	bool bResult = false;
	vector<int> ports;
	ports.push_back(12345);

	TestClassComplete *tcStub = new TestClassComplete_Stub(port, skeletonAddress, callbackAddress, rev, &bResult, ports);
	/*cout << "==== ENUM ====" << endl;
	Enum en;
	en = APPLE;
	Enum res = tcStub->handleEnum(en);
	cout << "Return value: " << res << endl;
	Enum resComp = tcStub->handleEnumComplex(en, 23, 'c');
	cout << "Return value: " << resComp << endl;
	vector<Enum> vecEn;
	vecEn.push_back(APPLE);
	vecEn.push_back(ORANGE);
	vecEn.push_back(APPLE);
	vecEn.push_back(GRAPE);
	vector<Enum> vecRes = tcStub->handleEnumArray(vecEn);
	for (Enum en : vecRes) {
		cout << "Enum value: " << en << endl;
	}
	vector<Enum> vecRes2 = tcStub->handleEnumList(vecEn);
	for (Enum en : vecRes2) {
		cout << "Enum value: " << en << endl;
	}
	vector<Enum> vecRes3 = tcStub->handleEnumComplex2(vecEn, 23, 'c');
	for (Enum en : vecRes3) {
		cout << "Enum value: " << en << endl;
	}


	cout << "==== STRUCT ====" << endl;
	Struct str;
	str.name = "Rahmadi";
	str.value = 0.123;
	str.year = 2016;
	Struct resStr = tcStub->handleStruct(str);
	cout << "Name: " << resStr.name << endl;
	cout << "Value:" << resStr.value << endl;
	cout << "Year" << resStr.year << endl;
	Struct resStr2 = tcStub->handleStructComplex(23, 'c', str);
	cout << "Name: " << resStr2.name << endl;
	cout << "Value:" << resStr2.value << endl;
	cout << "Year" << resStr2.year << endl;
	Struct str2;
	str2.name = "Trimananda";
	str2.value = 0.124;
	str2.year = 2017;
	vector<Struct> vecStr;
	vecStr.push_back(str);
	vecStr.push_back(str2);
	vector<Struct> vecRetStr = tcStub->handleStructArray(vecStr);
	for (Struct st : vecRetStr) {
		cout << "Name: " << st.name << endl;
		cout << "Value: " << st.value << endl;
		cout << "Year: " << st.year << endl;
	}
	vector<Struct> vecRetStr2 = tcStub->handleStructList(vecStr);
	for (Struct st : vecRetStr2) {
		cout << "Name: " << st.name << endl;
		cout << "Value: " << st.value << endl;
		cout << "Year: " << st.year << endl;
	}
	vector<Struct> vecRetStr3 = tcStub->handleStructComplex2(23, 'c', vecStr);
	for (Struct st : vecRetStr3) {
		cout << "Name: " << st.name << endl;
		cout << "Value:" << st.value << endl;
		cout << "Year" << st.year << endl;
	}
	vector<Enum> vecRes4 = tcStub->handleEnumStruct(vecEn, vecStr, 'c');
	for (Enum en : vecRes4) {
		cout << "Enum value: " << en << endl;
	}*/

	cout << "==== CALLBACK ====" << endl;
	CallBackInterface *cbSingle = new CallBack(2354);
	tcStub->registerCallback(cbSingle);
	cout << "Return value from callback: " << tcStub->callBack() << endl;
	/*CallBackInterface *cb1 = new CallBack(23);
	CallBackInterface *cb2 = new CallBack(33);
	CallBackInterface *cb3 = new CallBack(43);
	vector<CallBackInterface*> cb;
	cb.push_back(cb1);
	cb.push_back(cb2);
	cb.push_back(cb3);
	tcStub->registerCallbackArray(cb);
	cout << "Return value from callback: " << tcStub->callBack() << endl;*/
	/*CallBackInterface *cb4 = new CallBack(53);
	CallBackInterface *cb5 = new CallBack(63);
	CallBackInterface *cb6 = new CallBack(73);
	vector<CallBackInterface*> cblist;
	cblist.push_back(cb4);
	cblist.push_back(cb5);
	cblist.push_back(cb6);*/
//	tcStub->registerCallbackList(cblist);
//	cout << "Return value from callback: " << tcStub->callBack() << endl;
/*	tcStub->registerCallbackComplex(23, cblist, 0.1234);
	cout << "Return value from callback: " << tcStub->callBack() << endl;
	vector<Enum> vecRes5 = tcStub->handleAll(vecEn, vecStr, 'c', cblist);
	for (Enum en : vecRes5) {
		cout << "Enum value: " << en << endl;
	}

	vector<Enum> vecRes6 = tcStub->handleCallbackEnum(vecEn, 'c', cblist);
	for (Enum en : vecRes6) {
		cout << "Enum value: " << en << endl;
	}*/
	//vector<Enum> vecRes7 = tcStub->handleAllTwo(vecEn, vecStr, vecStr, vecEn, 'c', cblist, cblist);
	/*vector<Enum> vecRes7 = tcStub->handleAllTwo(vecEn, vecEn, vecStr, vecStr, 'c', cblist, cblist);
	for (Enum en : vecRes7) {
		cout << "Enum value: " << en << endl;
	}
	vector<Enum> vecRes8 = tcStub->handleEnumTwo(vecEn, vecEn);
	for (Enum en : vecRes8) {
		cout << "Enum value: " << en << endl;
	}
	vector<Enum> vecRes9 = tcStub->handleEnumThree(vecEn, vecEn, vecStr, vecStr);
	for (Enum en : vecRes9) {
		cout << "Enum value: " << en << endl;
	}
	vector<Struct> vecRetStr2 = tcStub->handleStructTwo(vecStr, vecStr);
	for (Struct st : vecRetStr2) {
		cout << "Name: " << st.name << endl;
		cout << "Value: " << st.value << endl;
		cout << "Year: " << st.year << endl;
	}
	vector<Struct> vecRetStr3 = tcStub->handleStructThree(vecStr, vecStr, vecStr);
	for (Struct st : vecRetStr3) {
		cout << "Name: " << st.name << endl;
		cout << "Value: " << st.value << endl;
		cout << "Year: " << st.year << endl;
	}*/

	return 0;
}

File: Sentinel/iotjava/iotrmi/Java/basics/TestClassAdvanced_Stub.java
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import iotruntime.master.CommunicationHandler;

public class TestClassAdvanced_Stub {

	public static void main(String[] args) throws Exception {

		CommunicationHandler comHan = new CommunicationHandler(true);
		int numOfPorts = 1;
		int[] ports = comHan.getCallbackPorts(numOfPorts);

		int port = 5010;
		//String address = "localhost";
		//String address = "192.168.2.191";	// RPi2
		//String skeletonAddress = "128.195.136.170";	// dc-9.calit2.uci.edu
		String skeletonAddress = "128.195.204.132";
		String callbackAddress = "128.195.204.132";	// dw-2.eecs.uci.edu (this machine)
		//String skeletonAddress = "192.168.2.108";	// RPi1
		//String callbackAddress = "192.168.2.191";	// RPi2
		int rev = 0;

		TestClassComplete_Stub tcstub = new TestClassComplete_Stub(port, skeletonAddress, callbackAddress, rev, ports);
		/*System.out.println("==== ENUM ====");
		Enum en = Enum.APPLE;
		Enum res = tcstub.handleEnum(en);
		System.out.println("Enum member: " + res);
		Enum resComp = tcstub.handleEnumComplex(en, 23, 'c');
		System.out.println("Enum member: " + resComp);

		Enum[] enArr = { Enum.APPLE, Enum.ORANGE, Enum.APPLE, Enum.GRAPE };
		Enum[] resArr = tcstub.handleEnumArray(enArr);
		System.out.println("Enum members: " + Arrays.toString(resArr));
		List<Enum> enArr2 = new ArrayList(Arrays.asList(enArr));
		List<Enum> resArr2 = tcstub.handleEnumList(enArr2);
		System.out.println("Enum members: " + resArr2.toString());
		Enum[] resArr3 = tcstub.handleEnumComplex2(enArr, 23, 'c');
		System.out.println("Enum members: " + Arrays.toString(resArr3));
		

		System.out.println("==== STRUCT ====");
		Struct str = new Struct();
		str.name = "Rahmadi";
		str.value = 0.123f;
		str.year = 2016;
		Struct strRes = tcstub.handleStruct(str);
		System.out.println("Name: " + strRes.name);
		System.out.println("Value: " + strRes.value);
		System.out.println("Year: " + strRes.year);
		Struct strRes2 = tcstub.handleStructComplex(23, 'c', str);
		System.out.println("Name: " + strRes2.name);
		System.out.println("Value: " + strRes2.value);
		System.out.println("Year: " + strRes2.year);
		Struct str2 = new Struct();
		str2.name = "Trimananda";
		str2.value = 0.124f;
		str2.year = 2017;
		Struct[] arrStr = { str, str2 };
		Struct[] arrRet = tcstub.handleStructArray(arrStr);
		for(Struct st : arrRet) {
			System.out.println("Name: " + st.name);
			System.out.println("Value: " + st.value);
			System.out.println("Year: " + st.year);
		}
		List<Struct> stList = new ArrayList(Arrays.asList(arrStr));
		List<Struct> stRetList = tcstub.handleStructList(stList);
		for(Struct st : stRetList) {
			System.out.println("Name: " + st.name);
			System.out.println("Value: " + st.value);
			System.out.println("Year: " + st.year);
		}
		List<Struct> stRetList2 = tcstub.handleStructComplex2(23, 'c', arrStr);
		for(Struct st : stRetList2) {
			System.out.println("Name: " + st.name);
			System.out.println("Value: " + st.value);
			System.out.println("Year: " + st.year);
		}
		Enum[] resArr3 = tcstub.handleEnumStruct(enArr, stList, 'c');
		System.out.println("Enum members: " + Arrays.toString(resArr3));*/

		System.out.println("==== CALLBACKS ====");
		CallBackInterface cbSingle = new CallBack(2354);
		tcstub.registerCallback(cbSingle);
		System.out.println("Return value from callback: " + tcstub.callBack());
		/*CallBackInterface cb1 = new CallBack(23);
		CallBackInterface cb2 = new CallBack(33);
		CallBackInterface cb3 = new CallBack(43);
		CallBackInterface[] cb = { cb1, cb2, cb3 };
		tcstub.registerCallbackArray(cb);
		System.out.println("Return value from callback: " + tcstub.callBack());
		List<CallBackInterface> cblist = new ArrayList<CallBackInterface>();
		CallBackInterface cb4 = new CallBack(53); cblist.add(cb4);
		CallBackInterface cb5 = new CallBack(63); cblist.add(cb5);
		CallBackInterface cb6 = new CallBack(73); cblist.add(cb6);
//		tcstub.registerCallbackList(cblist);
//		System.out.println("Return value from callback: " + tcstub.callBack());

		tcstub.registerCallbackComplex(23, cblist, 0.1234);
		System.out.println("Return value from callback: " + tcstub.callBack());
		Enum[] resArr4 = tcstub.handleAll(enArr, stList, 'c', cblist);
		System.out.println("Enum members: " + Arrays.toString(resArr4));

		Enum[] resArr5 = tcstub.handleCallbackEnum(enArr, 'c', cblist);
		System.out.println("Enum members: " + Arrays.toString(resArr5));
		//Enum[] resArr6 = tcstub.handleAllTwo(enArr, stList, stList, enArr, 'c', cblist, cblist);
//		Enum[] resArr6 = tcstub.handleAllTwo(enArr, enArr, stList, stList, 'c', cblist, cblist);
//		System.out.println("Enum members: " + Arrays.toString(resArr6));
		Enum[] resArr7 = tcstub.handleEnumTwo(enArr, enArr);
		System.out.println("Enum members: " + Arrays.toString(resArr7));
		Enum[] resArr8 = tcstub.handleEnumThree(enArr, enArr, stList, stList);
		System.out.println("Enum members: " + Arrays.toString(resArr8));
		List<Struct> stRetList2 = tcstub.handleStructTwo(stList, stList);
		for(Struct st : stRetList2) {
			System.out.println("Name: " + st.name);
			System.out.println("Value: " + st.value);
			System.out.println("Year: " + st.year);
		}
		List<Struct> stRetList3 = tcstub.handleStructThree(stList, stList, stList);
		for(Struct st : stRetList3) {
			System.out.println("Name: " + st.name);
			System.out.println("Value: " + st.value);
			System.out.println("Year: " + st.year);
		}*/
	}
}
